PooledQueueCache

jason-bragg opened the pull request:
This is the initial prototype of a pooled cache for the recoverable `EventHubStreamProvider`.

See: Recoverable Event Hub Persistent Stream Provider #1096
This is approximately what is described in "2. Introduce BigQueueCache..."

At this point, this queue cache is mainly a set of components that can be assembled into a fast in memory cache.  Further development will work to wrap and simplify these components.

The main goal of this cache is to support the caching of high throughput queues (~2k+ messages / queue / second) for an extended period of time (~30 minutes of data) amounting to gigs worth of data with minimal performance degradation due to garbage collection.

Unlike the simple queue cache, this queue purges data when cache capacity is reached, regardless of whether the events have been delivered or not.  The reasons for this is that it's intended for fast recoverable real time processing, so if processing gets too far behind (~15-30 minutes) it's preferable to drop those events and process new events in a timely manner than to fall further and further behind.

This implementation currently has no mechanism for regulating backpressure from stream processing delays.

Some components this implementation includes:
- A basic object pool
- A fixed size object pool that will trigger a purge of resources when it's capacity is reached.
- A fixed size buffer that allocates blocks of data used to cached serialized message payloads
- A block of tightly packed message metadata structures for cached messages.

Testing:
- Minimal BVT tests have been added - more are needed.
- GeneratorStreamProvider has been updated to use the PooledQueueCache.  It's test now exersize this code.

jason-bragg commented on src/OrleansProviders/Streams/Common/PooledCache/FixedSizeObjectPool.cs:
```diff
@@ -0,0 +1,59 @@
+
+        /// <summary>
+        /// Manages a memory pool of blockCount blocks of blockSize size.
+        /// Whenever a block is recycled, we call the purge passing the blocks buffer.
```
Common out of date

jason-bragg commented on src/OrleansProviders/Streams/Common/PooledCache/CachedMessagePool.cs:
```diff
@@ -0,0 +1,51 @@
+
+        /// <summary>
+        /// Creates a segment pool for the raw binary data and a cached message object
+        ///   pool for the messages.
```
Comments missleading

jason-bragg commented on src/OrleansProviders/Streams/Common/PooledCache/FixedSizeBuffer.cs:
```diff
@@ -0,0 +1,88 @@
+        ///   users of the block that the data is no longer valid.  The object passed is the buffer
+        ///   value used in any segment allocated by this block.  The pulled resource is this buffer.
+        ///   The implementation must release the pooled resource once they've cleaned up all segments
+        ///   referencing this block.
```
object passed to purge in now this block as an IDisposable.

gabikliot commented on src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs:
```diff
@@ -0,0 +1,295 @@
+            // if sequenceToken is null, iterate from newest message in cache
+            if (sequenceToken == null)
+            {
+                cursor.State = Cursor.States.Idle;
```
Why aren't we using constructors? Back to C?

gabikliot commented:
My biggest problem with this PR is the `PooledQueueCache` implementation: the `SetCursor`, `TryGetNextMessage `, and the `Cursor` class. Two reasons:

1) The code looks too familiar. This is since it is quite similar, but probably not identical, to [Simple Cache](https://github.com/dotnet/orleans/blob/master/src/OrleansProviders/Streams/Common/SimpleQueueCache.cs#L208) and to [KafkaCache](https://github.com/gigya/Orleans.KafkaStreamProvider/blob/master/src/Orleans.KafkaStreamProvider/KafkaQueue/TimedQueueCache/TimedQueueCache.cs#L324). We now have 3 pretty similar duplications of this pattern. Even if the pattern is golden (and it is not, as I will detail in point 2), code duplication is always bad. It basically means we did not think hard enough how to generalize the pattern to eliminate code duplication. **This part scares me the most** - any bug in this logic will need to be fixed in 3 places now. And as we keep duplicating this pattern to future caches, it only gets worse.

2) The pattern is definitely not golden. Although it does work, we discussed in the past a number of shorcomings of this pattern: its not intuitive ([wrong iterator logic](https://github.com/dotnet/orleans/issues/903#issuecomment-148588168)), hard to validate, and most importantly - it has a number of outstanding bugs we did not resolve yet - #1010, #1012, #903. So given the immaturity of this pattern, I would be very careful about duplicating it.

jason-bragg commented:
@Gabriel, I share some of your concerns regarding the cache, and have spent (and continue to spend) a fair amount of time on this logic (and as you know, I’ve not been the only one).  Despite these efforts a more intuitive, maintainable, generalized approach has not (yet) surfaced.

The various caches look similar and share a lot of code because their all based off the original 343 Industries implementation.  The SimpleCache was adapted to better suit the needs of AzureQueue which requires timely disposal of messages to ensure they don’t pop back into the queue.  The KafkaCache is based off of the simple cache and adjusted to hold data longer and introduced a time based purge.  This PooledCache is the closes to the 343 cache that is being used in production Halo services, and has the most generalizable purge characteristics. 

I am not yet convinced there will be a single generic caching architecture that will suit all needs.  Having said that, I suspect that a set of caching components (like those introduced in this PR) may serve as a cache developer’s toolbox, but this will need to be developed over time, as we (well at least I) don’t have the foresight to mastermind this from the start.

I won’t claim that there are no issues with this cache, only that there are no ‘known’ issues.  #1010, #1012 are not, to my knowledge, related to the cache (SimpleCache or otherwise), and issue #903 was introduced with the block tracking purge mechanic of the SimpleCache, which the PooledCached does not use.

jason-bragg commented:
@gabikliot 
Taking a step back from the technical details, this work is part of a port.  While a degree of refactoring is necessary my main goal is to port a production quality recoverable persistent stream provider into Orleans.  Once this is open sourced, any number of (smarter) engineers can refactor and improve the algorithms.  I'm very hesitant to rethink proven algorithms in a porting effort.

Let's get this ported and tested, then we can refactor and improve the implementations in a safer more focused manner.

gabikliot commented:
@jason-bragg, as long as you are committed to " then we can refactor and improve the implementations in a safer more focused manner" I am OK.

I did not do a thorough review of this code this time. I am well familiar with the design, block pool, etc.. and reviewed that in the past. Overall, I like the approach (except for the Cursor), so LGTM from me. 

If its ready from your side, I will merge.

jason-bragg commented:
@gabikliot Thanks for looking at this.
@jdom Is going to take a look at it too.  Trying to raise the truck number :).
I'll address any of his feedback, rebase, then it should be clear to merge.

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/FixedSizeBuffer.cs:
```diff
@@ -0,0 +1,85 @@
+namespace Orleans.Providers.Streams.Common
+{
+    /// <summary>
+    /// Manages a contiguous block of memory. Calls purge with itself when it is purged.
```
This comment is a little weird. I think I understand what you mean, but you might want to rephrase a little bit. Maybe the confusion is from the method names you chose. Perhaps RequestPurge and OnPurgeRequested (or something around those lines) would make it self explanatory?

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/FixedSizeObjectPool.cs:
```diff
@@ -0,0 +1,60 @@
+        public override T Allocate()
+        {
+            T obj;
+            lock (locker)
```
probably can do without a lock and use `Interlocked.CompareExchange` instead, or do you actually want to run the purge logic inside the lock too? I would think we don't want to do that, but you are probably more familiar with this

jason-bragg replied:
The signaling of the purge does not necessarily run the purge logic at that time.  It's implementation specific, but the signaling of the purge, the allocation, and tracking of the allocated block needs to all occur as a single thread safe state change.

jason-bragg replied:
Need to lock free as well though, as ObjectPool is not thread safe.  Will add.

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/FixedSizeObjectPool.cs:
```diff
@@ -0,0 +1,60 @@
+                obj = base.Allocate();
+
+                // track used objects
+                usedObjects.Enqueue(obj);
```
this puts the object in both the base ObjectPool stack and on a queue? is there a benefit to have both collections?

jason-bragg replied:
Not sure what you mean.  Allocate removes it from the base state, as the base stack is a collection of available resources.  The usedObjects collection is a set of allocated resources, and is used to signal that resources need to be freed (returned to the pool to be reused) when they exceed the specified number of max allocated resources.  In most cases a resource will either be in the base stack of available resources or in the usedObjects collection of allocated resources, but shouldn't ever be in both.

jdom replied:
yeah, I found that after reading more, sorry

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/IObjectPool.cs:
```diff
@@ -0,0 +1,60 @@
+        /// Notifies the object that it has been purged, so it can reset itself to
+        ///   the state of a newly allocated object.
+        /// </summary>
+        public virtual void OnDispose()
```
Maybe `OnResetState`? OnDispose gives the impression that you should just free unmanaged resources, but not necessarily prepare the object for re-use.

jason-bragg replied:
+1

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/ObjectPool.cs:
```diff
@@ -0,0 +1,44 @@
+
+        public void Free(T resource)
+        {
+            pool.Push(resource);
```
could this be dangerous if we NEVER completely release a pooled resource? I thought the idea is to use individual PooledResources (which they contain several cached items) to reduce GC preassure, but here you are also never releasing the PooledResources themselves... basically by not allowing the GC to even collect the PooledResource... is that by design?

jdom replied:
OK, I talked with Jason in person, and it is in fact doing 2 layers of reducing GC pressure, and that was perf tested a lot in 343. That part might be fine, but I would still like to put a cap on the maximum number of PooledResources that would be stored in the stack waiting for re-allocation. Under normal load I wouldn't expect to have thousands of un-allocated objects standing by... I would prefer the GC to reclaim them... I would only expect a few of them at most (maybe 5 or 10?) waiting for reallocation. Still putting a cap that is bigger than that (50? 100?) I would consider safer.

jason-bragg replied:
>"not allowing the GC to even collect the PooledResource... is that by design"

Yes.  We allocate relatively few large objects that contain few object references and never free them.  The fact that there are relatively few objects with few references makes it cheap for garbage collection to check to see if they need collected, and the fact that we never free them means that they will eventually settle in the garbage collector and no longer (or rarely) be moved around or trigger other data to be moved during compaction.

Is all of this necessary?  Unclear.  We've not tested the large objects without the object pools to see what the perf hit is, but we know that keeping the objects in pools is cheaper than letting the garbage collection constantly compact and collect them, we just don't know how much cheaper it is.

jason-bragg replied:
>"I would only expect a few of them at most (maybe 5 or 10?) waiting for reallocation."

It is true that this cache mechanism, over time, may allocate more objects than are needed, but there is no indication so far that this number of unallocated blocks is significant.  In the original 343 code, we had more logging and performance counters than what has been ported.  These were not ported because the mechanisms used for this were proprietary.  In future passes, logging and performance metrics using Orleans tooling will be introduced, and if this logging indicates a need for caps, we should add them.  At this stage of development, I’m reluctant to add complexity based on speculation.

With this PR I'll open an issue to add logging and performance metrics to pooled queue cache.

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs:
```diff
@@ -0,0 +1,306 @@
+
+namespace Orleans.Providers.Streams.Common
+{
+    public class PooledQueueCache<TQueueMessage, TCachedMessage>
```
Description of what this class does? Due to the double layer of pooling and also different objects for metadata and the message payload, it is a little confusing to follow (at least in the diff). Would it be possible to have a place somewhere that puts all the pieces together? Maybe in ObjectPool which I assume should be the "root" of all this hierarchy? Or if there's a better place, go for it

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs:
```diff
@@ -0,0 +1,306 @@
+            uint purgeCount = 0;
+
+            // If there are purge requests, perform purges
+            if (!purgeQueue.IsEmpty)
```
no need to check for empty first, just try to dequeue

jason-bragg replied:
+1

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs:
```diff
@@ -0,0 +1,306 @@
+
+        private uint PerformBlockPurge(IDisposable purgeRequest)
+        {
+            uint purgeCount = 0;
```
any reason to count? seems you are completely ignoring this

jason-bragg replied:
It was being used for logging and backpressure, neither of which exist in the port, so I'll removed.

jdom commented on src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs:
```diff
@@ -0,0 +1,306 @@
+                if (messageBlocks.Last.Value.IsEmpty && !messageBlocks.Last.Value.HasCapacity)
+                {
+                    messageBlocks.Last.Value.Dispose();
+                    messageBlocks.RemoveLast();
```
I understand this is probably not called from multiple threads, but still for readability it might be good to store `messageBlocks.Last.Value` in a variable and then do `Remove()`, `IsEmpty`,  `HasCapacity` and `Dispose()` on it

jdom commented on src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs:
```diff
@@ -0,0 +1,197 @@
+            this.bufferPool = bufferPool;
+            operations = new CacheDataAdapter(bufferPool, disposable => cache.Purge(disposable));
+            cache = new PooledQueueCache<GeneratedBatchContainer, CachedMessage>(operations);
+        }
```
albeit safe, I'd initialize the `cache` field before the `operations` field, as it uses the cache object in its closure. FxCop would even complain that you are passing a reference to `this` before the constructor finished, so even more correct would be to initialize a `queueCache` variable and use that in the closure (as well as assign that same value to the `cache` field).

jason-bragg replied:
+1

jdom commented on src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs:
```diff
@@ -0,0 +1,197 @@
+            private ArraySegment<byte> SerializeMessageIntoPooledSegment(GeneratedBatchContainer queueMessage)
+            {
+                // serialize payload
+                byte[] serializedPayload = SerializationManager.SerializeToByteArray(queueMessage.Payload);
```
would this be improved in a non-test version? because if you have to always use a temporary new byte[], then the GC will fragment uncessarilly, albeit objects never leave gen0... Can you think of a way to avoid this allocation entirely (or is it worth it)?

jdom replied:
or even allocate from one pool of big buffer arrays, and then move it to a smaller pool if needed... Really not sure how worth it this is, given that it never leaves gen0 anyway

jason-bragg replied:
This is dependent on queue data type.  If the queue data payload is available as raw binary data like in Kafka or EventHub this is not necessary, but in this case, the generated generates fully populated objects, so we need to serialize them.  In any case, this cost is minimal.  Gen0 allocations are, unless excessive, insignificant.  Given all the other overhead, I doubt optimizing this would be noticeable.

jdom commented on src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs:
```diff
@@ -0,0 +1,197 @@
+            {
+                //Deserialize payload
+                var stream = new BinaryTokenStreamReader(cachedMessage.Payload.ToArray());
+                object payloadObject = SerializationManager.Deserialize(stream);
```
use the BinaryTokenStreamReader constructor overload that takes a list of ArraySegment (or even create an overload that takes just 1 segment), instead of using the Enumerable.ToArray() method

jason-bragg replied:
+1

jdom commented on src/Tester/TestStreamProviders/Generator/GeneratorPooledCache.cs:
```diff
@@ -0,0 +1,197 @@
+                var stream = new BinaryTokenStreamReader(cachedMessage.Payload.ToArray());
+                object payloadObject = SerializationManager.Deserialize(stream);
+                return new GeneratedBatchContainer(cachedMessage.StreamGuid, cachedMessage.StreamNamespace,
+                    payloadObject, new EventSequenceToken(cachedMessage.SequenceNumber));
```
reuse `GetSequenceToken` method instead?

jason-bragg replied:
This would introduce extra call + cast.  GeneratedBatchContainer expects EventSequenceToken.

jason-bragg replied:
As a side note, these two sequence tokens happen to be the same for this implementation, but they will not always be the same.  The sequence token in the batch is the token to be delivered to the application, while the sequence token returned by GetSequenceToken is used for ordering in the cache, which represents the order of data in the queue.  These may not always be the same.  Consider a queue that has duplicates of application layer data.

Application sends event 1,2,3 gets a timeout on 3, and resends.  The first send of 3 that timed out actually succeeded, so the queue has application layer data 1,2,3,3, but queue ordering of 1,2,3,4.  The sequence token provided to the cache must reflect queue ordering (1,2,3,4), but when sending to the application, which will perform deduplication base off the sequence token, tokens should be 1,2,3,3 so that the second 3 will be detected as a duplicate and ignored.

amamh commented:
@jason-bragg 

Sorry if this is not the right place to ask. What is the benefit of this caching system compared to using a simple cache with an internal linked list? Why do we have a "pool" of objects?

Is there another place to ask these questions?

jason-bragg commented:
@amamh 
The .net garbage collector keeps objects in 'generations'.  When an object is initially allocated, it is generation 0 (gen0).  When garbage collection is run on gen0, objects that remain are gen1.  When garbage collection is run on gen1, objects that remain are placed in gen2.  When garbage collection is run on gen2, remaining objects are compacted to reduce memory fragmentation.  As far as cpu cost of garbage collection, the cheapest is to clean up an object in gen0, then gen1, then gen2 in first run, with gen2 after compaction as the most expensive.

With the above in mind, if a stream provider is caching messages long enough for them to get into gen2 and become compacted, the cost of garbage collection will be high.  If the stream provider is burning through a large number of messages, this can easily result in garbage collection taking up a significant portion of the available cpu.

To avoid this, this pooled cache allocates memory and never releases it.  It keeps freed resources available in pools that remain in application use through the life of the service.  This means these objects go to gen2, are compacted, and then stay there.  This is relatively cheap, as the only cost they now incur is the cost of checking to see if they should be freed in each collection cycle.  Since this cache uses small numbers of large objects with relatively simple object graphs, they are less costly to check then large numbers of smaller objects with more complex object graphs.

This architecture reduces GC cost, but introduces a fair amount of complexity, so I’d not suggest using it if simpler solutions suffice.  It really depends on the performance needs of your service.  The requirements that necessitated a solution of this sort were a stream provider with throughput of 2k+ events per second with a 30-minute cache (about 2 gig of messages).  With 2 gig of messages compacted into gen2 and 2000 of the oldest of them being removed from gen2 every second, the garbage collection was preventing us from hitting our perf targets.

amamh commented:
@jason-bragg 

Thanks, that's very interesting. I really appreciate your taking the time to explain this :)

I have some questions, if you don't mind:

>  this pooled cache allocates memory and never releases it

Does that mean I cannot leave the server running for days? if the memory is never released, **at some point that will cause an issue**, isn't it?

I have made a persistent stream using your cache and another one using a simple linked list cache which also never purges. I then wrote a **simple benchmark test where I send 1 million int messages on one stream and measure how long it takes to send/receive every 10k messages**.

The one with the [simple cache](https://github.com/amamh/orleans-PipeStreamProvider) performs 6 times faster when writing/sending and about 30% faster when reading/receiving than the [one with the pooled cache](https://github.com/amamh/orleans-FastPipeStreamProvider).

I'm not expecting you to look into these two project, but maybe you can suggest **a test scenario where the pooled cache should perform better**?

jason-bragg commented:
>Does that mean I cannot leave the server running for days? if the memory is never released, at some point that will cause an issue, isn't it?

The memory allocated in the pool is never freed, but it is capped.  If one configured the pooled cache to use 1 gb of memory, it will store 1 gb worth of messages, then, as new messages are read from the queue, the oldest messages will be purged from the cache.  While the messages are purged, the memory they were kept in is returned to the pools to be reused for the next set of messages being added to the cache.  So while the memory is never freed, it does not grow indefinitely.

>using a simple linked list cache which also never purges.

To be clear, the pooled cache does purge messages, it just does not release the resources to be collected by GC.  It pools them for reuse.

>The one with the simple cache performs 6 times faster when writing/sending and about 30% faster when reading/receiving than the one with the pooled cache.

Not sure how to make sense out of this, as the cache should have no bearing on the send speed.

>maybe you can suggest a test scenario where the pooled cache should perform better?

1) You'll want to add perf counter "%time spend in gc" to your monitoring, as this is what we're optimizing.  If your linked list cache is not big enough and kept around long enough to create GC pressure, then you really don't have a problem to solve.  https://msdn.microsoft.com/en-us/library/w8f5kw2e(v=vs.110).aspx
2) Try to figure out how many messages it would take to chew up about 1gb of memory.  I don't know your message size, so you'll need to do the math on this.
3) Configure your linked list cache to store the number of messages calculated in 2, and when full free up old messages.
4) Run your test sending at least 4 times the number of messages through the stream as your cache size (message count calculated in 2).

The above should create a situation where the %time spend in GC should increase once your cache is full.
Measure the events per second sent through the system during the run.  I'd target around 2k+ per second, but I'm not sure what your queue can handle.  If you have low throughput, gc may not be an issue.

If you see an increase in %time in gc in the above run, configure the pooled cache to have the same 1gb size limit (I suggest using 1024 1mb buffers) and run the same test using the pulled cache.  What I'd expect is to have lower %time in gc and better performance.

amamh commented:
> While the messages are purged, the memory they were kept in is returned to the pools to be reused for the next set of messages being added to the cache. So while the memory is never freed, it does not grow indefinitely.

So you basically mark that region as writable and **overwrite** whatever is there with new messages? That's the only way I can think of that would allow to store new messages without the old ones getting GC'd.

> Not sure how to make sense out of this, as the cache should have no bearing on the send speed.

I guess the reason is because I'm running all on the same machine so the caching system takes some CPU power which slows down writing. I shouldn't have benchmarked writing at all.



Thanks for suggesting a testing method.

jason-bragg commented:
>So you basically mark that region as writable and overwrite whatever is there with new messages?

Something like that.  Imagine you have 10 million objects each of which has an int and a long that you want to cache.  Rather than keeping around those 10 million objects you could keep around 1000 arrays of 10000 structures that each have an int and a long.  When a new object comes in, you write its values to into a spot in one of the structures in one of the arrays, then release the object.  Since you released the object quickly it is cleaned up in gen0.  If you need the object some 20 minutes later you can allocate a new one and copy the data from the struct, use the object, then free it.  Also a gen0 operation.  Since the arrays never get cleaned up, they never really incur a collection cost.  So now you can keep 10 million objects in memory for any amount of time with no gc hit, at the cost of copying the objects in and out of the memory you keep around for this purpose.  This cache is not exactly like that, but pretty close.

Where I think your original test may have been insufficient is that it never freed up the messages from the queue.  It's the cost of cleaning up old messages that is the biggest hit, so by avoiding that there is far less difference between the two caches.

I've not perf tested this implementation of the cache yet, but in previous versions I found it simpler to write a console app that just pumped data into the cache directly rather than going through the full streaming architecture.  It made it easier to identify bottlenecks, iterate, and push the cache to its limits.  In those tests I was able to vet the perf with a cache up to 16gb.  Not suggesting you do this, it's your call, but for me, it did simplify the targeted testing of the cache behavior.

jdom commented:
code looks good to me. There are several typos in the comments, not sure if it's worth highlighting

jason-bragg commented:
KK, I still have another push, so i'll take another look at the comments.

jason-bragg commented:
Bad rebase, plz don't merge. Fixing

jason-bragg commented:
Fixed.  Please sanity check prior to merge.
